bodo/pandas/physical/expression.h

```diff
@@ -487,3 +461,155 @@ class PhysicalConjunctionExpression : public PhysicalExpression {
+    }
+
+   protected:
+    std::string comparator;
+};
+
+/**
+ * @brief Physical expression tree node type for binary op non-boolean arrays.
+ *
+ */
+class PhysicalBinaryExpression : public PhysicalExpression {
+   public:
+    PhysicalBinaryExpression(std::shared_ptr<PhysicalExpression> left,
+                             std::shared_ptr<PhysicalExpression> right,
+                             duckdb::ExpressionType etype) {
+        children.push_back(left);
+        children.push_back(right);
+        switch (etype) {
+            default:
+                throw std::runtime_error("Unhandled binary expression type.");
+        }
+    }
+
+    virtual ~PhysicalBinaryExpression() = default;
+
+    /**
+     * @brief How to process this expression tree node.
+     *
+     */
+    virtual std::shared_ptr<ExprResult> ProcessBatch(
+        std::shared_ptr<table_info> input_batch) {
+        // We know we have two children so process them first.
+        std::shared_ptr<ExprResult> left_res =
+            children[0]->ProcessBatch(input_batch);
+        std::shared_ptr<ExprResult> right_res =
+            children[1]->ProcessBatch(input_batch);
+        // Try to convert the results of our children into array
+        // or scalar results to see which one they are.
+        std::shared_ptr<ArrayExprResult> left_as_array =
+            std::dynamic_pointer_cast<ArrayExprResult>(left_res);
+        std::shared_ptr<ScalarExprResult> left_as_scalar =
+            std::dynamic_pointer_cast<ScalarExprResult>(left_res);
+        std::shared_ptr<ArrayExprResult> right_as_array =
+            std::dynamic_pointer_cast<ArrayExprResult>(right_res);
+        std::shared_ptr<ScalarExprResult> right_as_scalar =
+            std::dynamic_pointer_cast<ScalarExprResult>(right_res);
+
+        arrow::Datum src1;
+        if (left_as_array) {
+            src1 = arrow::Datum(prepare_arrow_compute(left_as_array->result));
+        } else if (left_as_scalar) {
+            src1 =
+                arrow::MakeScalar(prepare_arrow_compute(left_as_scalar->result)
+                                      ->GetScalar(0)
+                                      .ValueOrDie());
+        } else {
+            throw std::runtime_error(
+                "PhysicalBinaryExpression left is neither array nor scalar.");
+        }
+
+        arrow::Datum src2;
+        if (right_as_array) {
+            src2 = arrow::Datum(prepare_arrow_compute(right_as_array->result));
+        } else if (right_as_scalar) {
+            src2 =
+                arrow::MakeScalar(prepare_arrow_compute(right_as_scalar->result)
+                                      ->GetScalar(0)
+                                      .ValueOrDie());
+        } else {
+            throw std::runtime_error(
+                "PhysicalBinaryExpression right is neither array nor scalar.");
+        }
+
+        arrow::Result<arrow::Datum> cmp_res =
+            arrow::compute::CallFunction(comparator, {src1, src2});
```

Comment:
Would be great if we could refactor some of the duplicated Arrow compute call code across these classes.

Reply:
done.